File: client/src/types/recommendation.ts
import type { University } from '@/types/university'

export interface RecommendationItem {
  university: University
  score: number
  ml_score: number
  match: MatchResult
  cost: CostEstimate
  rank: number
  category: 'reach' | 'target' | 'safety'
}

export type MatchCriterionStatus = 'available' | 'unknown' | 'not_applicable'
export interface MatchCriterion { status: MatchCriterionStatus; score: number | null; weight: number; reason: string }
export interface MatchResult { match_score: number | null; coverage: number; available_criteria: number; total_criteria: number; breakdown: Record<'gpa' | 'test' | 'budget' | 'program', MatchCriterion> }
export interface CostEstimate { tuition: number | null; living_cost: number | null; application_fee: number | null; potential_aid: number | null; estimated_total: number | null; currency: string | null; missing_fields: string[] }
export interface SimulationInput { gpa?: number; gre_score?: number; budget?: number; budget_currency?: string; top_k?: number }

export interface RecommendationList {
  generation_id: string | null
  model_version: string | null
  generated_at: string | null
  recommendations: RecommendationItem[]
}

File: server/app/schemas/recommendation.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.university import UniversityResponse


class MatchCriterionResponse(BaseModel):
    status: Literal["available", "unknown", "not_applicable"]
    score: float | None
    weight: float
    reason: str


class MatchResultResponse(BaseModel):
    match_score: float | None
    coverage: float
    available_criteria: int
    total_criteria: int
    breakdown: dict[str, MatchCriterionResponse]


class CostEstimateResponse(BaseModel):
    tuition: float | None
    living_cost: float | None
    application_fee: float | None
    potential_aid: float | None
    estimated_total: float | None
    currency: str | None
    missing_fields: list[str]


class RecommendationItem(BaseModel):
    university: UniversityResponse
    score: float
    ml_score: float
    match: MatchResultResponse
    cost: CostEstimateResponse
    rank: int
    category: str


class RecommendationList(BaseModel):
    generation_id: str | None
    model_version: str | None
    generated_at: datetime | None
    recommendations: list[RecommendationItem]


class SimulationRequest(BaseModel):
    gpa: float | None = Field(default=None, ge=0, le=4)
    gre_score: int | None = Field(default=None, ge=260, le=340)
    budget: float | None = Field(default=None, ge=0)
    budget_currency: str | None = Field(default=None, min_length=3, max_length=3)
    top_k: int = Field(default=10, ge=1, le=25)

    model_config = {"extra": "forbid"}

    @field_validator("budget_currency")
    @classmethod
    def normalize_currency(cls, value: str | None) -> str | None:
        return value.upper() if value else None
